presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyRowExpressions.java

```diff
@@ -0,0 +1,132 @@
+    {
+        super(new Rewriter(metadata));
+    }
+
+    private static class Rewriter
+            implements PlanRowExpressionRewriter
+    {
+        private final RowExpressionOptimizer optimizer;
+        private final LogicalExpressionRewriter logicalExpressionRewriter;
+
+        public Rewriter(Metadata metadata)
+        {
+            requireNonNull(metadata, "metadata is null");
+            this.optimizer = new RowExpressionOptimizer(metadata);
+            this.logicalExpressionRewriter = new LogicalExpressionRewriter(metadata.getFunctionManager());
+        }
+
+        @Override
+        public RowExpression rewrite(RowExpression expression, Rule.Context context)
+        {
+            return rewrite(expression, context.getSession().toConnectorSession());
+        }
+
+        private RowExpression rewrite(RowExpression expression, ConnectorSession session)
+        {
+            RowExpression optimized = optimizer.optimize(expression, SERIALIZABLE, session);
+            if (optimized instanceof ConstantExpression || !BooleanType.BOOLEAN.equals(optimized.getType())) {
+                return optimized;
+            }
+            return RowExpressionTreeRewriter.rewriteWith(logicalExpressionRewriter, optimized, true);
+        }
+    }
+
+    @VisibleForTesting
+    public static RowExpression rewrite(RowExpression expression, Metadata metadata, ConnectorSession session)
+    {
+        return new Rewriter(metadata).rewrite(expression, session);
+    }
+
+    private static class LogicalExpressionRewriter
+            extends RowExpressionRewriter<Boolean>
+    {
+        private final FunctionResolution functionResolution;
+        private final LogicalRowExpressions logicalRowExpressions;
+
+        public LogicalExpressionRewriter(FunctionManager functionManager)
+        {
+            requireNonNull(functionManager, "functionManager is null");
+            this.functionResolution = new FunctionResolution(functionManager);
+            this.logicalRowExpressions = new LogicalRowExpressions(new RowExpressionDeterminismEvaluator(functionManager), new FunctionResolution(functionManager), functionManager);
+        }
+
+        @Override
+        public RowExpression rewriteCall(CallExpression node, Boolean isRoot, RowExpressionTreeRewriter<Boolean> treeRewriter)
+        {
+            if (BooleanType.BOOLEAN.equals(node.getType()) && functionResolution.isNotFunction(node.getFunctionHandle())) {
+                return rewriteBooleanExpression(node, isRoot);
+            }
+            if (isRoot) {
+                return treeRewriter.rewrite(node, false);
+            }
+            return null;
+        }
+
+        @Override
+        public RowExpression rewriteSpecialForm(SpecialFormExpression node, Boolean isRoot, RowExpressionTreeRewriter<Boolean> treeRewriter)
+        {
+            if (BooleanType.BOOLEAN.equals(node.getType()) && isConjunctiveDisjunctive(node.getForm())) {
```

Comment:
`BooleanType.BOOLEAN.equals(node.getType())` should be an assertion instead of a condition. Put a checkState